### DISCUSSION TITLE
key missmatch error.

### DISCUSSION THREAD

**Discussion by pnunn** [rustdesk/rustdesk#1418] [Q&A]
Hi Guys,

I've just set up a rustdesk server with docker and have enabled forceing encryption with the _k options, but now, no matter what I try and connect too (weather the client actually exists or not) I get an immediate key missmatch error.

I've looked in the hbbs container at the id_ed25519.pub key there and despite the fact that it seems oddly long for a 25519 key it doesn't want to work at all.

The key when I cat it is 

idsomelongrandomnumber=root@somerandomnumber:~#

My understanding is that the bit I should be putting in the client is idsomelongrandomnumber= and not the root@somerandomnumber:~# but it doesn't matter if I use the first string, or the entire string I still get the same error when I try and connect.

Am I doing something wrong here?

**Comment by coolboole** [comment]
I found the solution to thjis in #499 under [this comment](https://github.com/rustdesk/rustdesk/issues/499#issuecomment-1160578895)

**Reply by pnunn** [reply]
Thanks @coolboole. In my setup one of the . directories was empty, so I put the keys from the other one into it and the key miss match error went away (I think it actually went when I put the same key after the _k in the docker file as well.

Not the connection starts, I can see the relay connection in the log files, but the connection just times out regardless of what machine I try and connect too. There is nothing in the logs and I can't see any return traffic from the rust server in the firewall trace either, the traffic just seems to vanish.

No idea where to go from here.

**Comment by shaojun** [comment]
@pnunn any luck on this issue?
I'm running the open source rustdesk server via the official `compose.yml`, and i believe everything is there:
```
root@ubuntu-s-1vcpu-1gb-sgp1-01:~/rustdesk# ls
compose.yml  data
root@ubuntu-s-1vcpu-1gb-sgp1-01:~/rustdesk# ls data/
db_v2.sqlite3  db_v2.sqlite3-shm  db_v2.sqlite3-wal  id_ed25519  id_ed25519.pub
```
and I set my server's public domain name into rustdesk client(`Version: 1.2.3-2`):
![image](https://github.com/rustdesk/rustdesk/assets/3241829/18b2d47e-6f54-49a7-ac45-e079d8cfefc2)

then, the same as yours, no matter if the target client exists or not, once i click connect, the `Key mismatch` poped up:
![image](https://github.com/rustdesk/rustdesk/assets/3241829/f31fae56-c192-400e-95d9-be0d486ed304)

**Reply by Xitee1** [reply]
Just got the same issue. Probably the update of the rustdesk server I did yesterday causes it.

**Reply by Xitee1** [reply]
Downgrading my server back to 1.1.9 fixed the issue for now.
Probably this causes the issue for me:
From changelog in v1.1.11-1: `This version changes default -k of hbbs to -, which means setting key on client side is mandatory. If you want to run empty -k as before, please run hbbs -k "".`

**Reply by shaojun** [reply]
i finally overcome this error by manually input the public key.
so do you plan to revert this change or keep it? otherwise, I think the document has to be updated.

**Reply by Xitee1** [reply]
> so do you plan to revert this change or keep it? otherwise, I think the document has to be updated.

I'm not responsible for this change nor a maintainer of RustDesk. Just found this GitHub issue by looking up the error I've got.

**Comment by mingzhao2019** [comment]
I changed to docker image tag with 1.1.9 and it solved the issue...

**Comment by coderunway** [comment]
I reverted back to 1.1.10-13 and it solved .  Is this a bug in latest release ? Is there a tracking Issue for this bug , which will fix this issue ?

**Comment by brunocardosobc** [comment]
I had the same "key mismatch" error and I solved this way:

On the docker-server folder of my docker, I've get the id_ed25519.pub content:
![image](https://github.com/user-attachments/assets/a4d18f1b-5fc5-4d67-8fde-af8a9d6e16b0)

That is the public key you need to fill in the Key field of Network Settings of your client:
![image](https://github.com/user-attachments/assets/957ab170-49b1-43c0-85c0-4041de08d459)

That's all!

**Reply by drd2aiki** [reply]
which docker tutorial did you use to install this? having the same issue with key mismatch. copying the ID isn't fixing it.

**Reply by mattbrac3** [reply]
I am also getting the key mismatch error as well and I've copied and pasted the Public key in the places you mentioned.

**Reply by zzztazmanzzz** [reply]
I want to make sure everyone sees the key and this one and mine ended in a = sign.  This usually would have been a delimiter but in this case its part of the key that is required.

**Comment by Anas-Alhariri** [comment]
I found a fix.

Steps:
1. rm -f /opt/rustdesk/id_ed25519*
2. ssh-keygen -t ed25519 -f /opt/rustdesk/id_ed25519 -N ""
3. restart the rustdesk server components:
systemctl restart rustdeskrelay.service
systemctl restart rustdesksignal.service

Done! I hope that helps.

**Reply by Troubladore** [reply]
@Anas-Alhariri Thank you for posting this. I see how running this gens a new key in the folder where the service reads it from on startup. I was wondering whether you needed the public key value generated by this command in your client Network->ID/Relay server->key setting, and whether you were using a docker compose setup, and if so, what you used for , and what you were using for services:hbbs:command: hbbs -r <server:port> -k ?

Sorry for all the questions, but I haven't been able to get this working with compose and the latest release

**Reply by Anas-Alhariri** [reply]
@Troubladore 
I am using rustdesk server in linux based container image inside my Proxmox VE Server.

If you have installed RustServer inside Dcoker container. then you should follow those steps inside the docker container. If that didn't work, then you might need to copy those same files into the linux server to `/opt/rustdesk/` directory (most likely won't have this directory in this case, so you will need to create the rustdesk folder there before you copy the files into it)

plus, docker containers do NOT user systemctl to control the services, so instead of the second commands in my previous post, you should try this:
2. restart the rustdesk server service:
`service rustdeskrelay.service restart`
`service rustdesksignal.service restart`

**Note: Since you are using docker container for rustdesk, You might need to re-start the rustdesk services mentioned above after every start of your container to keep rustdesk running.

Let me know how it goes and if you need further assistant.

If you need further help. you may send me a message on eng.anas.alhariri@gmail.com to look into your specific environment.

I don't work for rust desk, but I would love to help. 

I hope this helps.

Cheers!

**Reply by Troubladore** [reply]
Thanks @Anas-Alhariri!

With the docker setup I'm using, the keys end up being stored in /opt/rustdesk/data. Taking down the containers, deleting the keys, and starting the containers again led to the keys being automatically regenerated (I didn't have to manually create them). Then following the current docs on client-side configuration (stopping the client service, then under Network Settings only setting up the ID w/o relay and adding the new public key, then restarting the client service) did allow me to connect. Besides these steps, having this command config in config.yml seemed necessary:
`services:
    ...
    hbbs:
        ...
        command: hbbs -r <my.server.dns>:<port> -k _
        ....
 `

Note /opt/rustdesk/data folder may not be a standard configuration for Pro users.

**Reply by Anas-Alhariri** [reply]
Yes that would also force the files to be regenerated.
When using a self-hosted RustDesk Server, you need to add the IP Address/hostname of the device in the ID Server field in the settings of each RustDesk client you want to use to access from or to.

This will also help the discovery process of the devices that use the same ID Server.

if you want to make sure that you will be able to access all those devices remotely outside of your local network. using your Public IP (if you have a static IP) will allow you to do so, but make sure to open the UDP and TCP ports required as explained in RustServer deployment documentation.

Alternatively, you make have a container running with no-ip service on it to give you a free domain name [something like: {yourChoosenName}.myddns.me

This no-ip service will keep the domain name linked to the up-to-date public ip address if you don't have a static one.

Whether you used IP or domain name, you will need to make sure that ip/domain name is used in the ID Server field.

I hope that helped :)

**Comment by vakus** [comment]
Putting this here if anyone encounters this issue with the same root cause as mine...

The location of the `compose.yml` file I was using was inside of my user directory. This was working fine when I started the services manually either via `docker compose up -d` or `docker restart hbbs`, but it would break after every restart of the system.

The issue was that my home directory was set to be encrypted - so at the time the docker starts with OS start, the user directory was **not there yet**.

The solution I took, I moved the compose and data files outside of my encrypted directory, allowing docker to access those files on system start.

**Comment by sudheer-raja** [comment]
Also, adding my solution just in case if anybody comes across the same issue. 

I didn't make changes to the Keys or Rustdesk configuration so it didn't really make sense to try all the other solutions so I tried something very simple. 

1. I stopped the running _hbbs_ and _hbbr_ services.
```
docker stop hbbs hbbr
```
2. Removed both the containers
```
docker rm hbbr hbbs
```
3. Then pulled the latest images. I am using docker compose so used the following command/
```
# pull new images to get the latest
docker compose pull
```
4. Then recreated the containers
```
# recreate containers
docker compose up -d
```

This fixed my issue. I could connect to my remote machines using the relay I set up.

**Comment by tedtieken** [comment]
Still happening.  Synology with docker with tailscale.  Solution for me was to pin the docker image to 1.1.10-3

```yaml
services:
  hbbs:
    container_name: hbbs
    image: rustdesk/rustdesk-server:1.1.10-3   # was :latest
    command: hbbs
    ...
  hbbr:
    container_name: hbbr
    image: rustdesk/rustdesk-server:1.1.10-3   # was :latest
    command: hbbr  
    ...
```

this appears to be this issue:  https://github.com/rustdesk/rustdesk-server/issues/590
